### ISSUE TITLE
Write JavaScript Bindings

### ISSUE THREAD

**Issue by TannerRogalsky** [TannerRogalsky/love.js#27]
JS bindings into liblove would, I think, make this the best 2D JS game engine. Could the Lua bindings setup be used in such a way as to quickly generate JS bindings?

**Comment by kripken** [comment]
Might be worth looking into the emscripten WebIDL binder, if you want to create bindings that look like JS objects, from C++ objects. Or if these are C bindings, something simpler could be done, just direct calls would work for most things.

**Comment by TannerRogalsky** [comment]
@kripken Yeah, Love is, perhaps, unique because it already has it's C++ functionality exposed as Lua bindings. I need to look into if that infrastructure can be leveraged for the JS bindings. Otherwise, maintaining parity between JS and Lua bindings will be a large undertaking.

**Comment by Positive07** [comment]
If it could be something like FFI, where you require a library and get an interface to JS, I would love it! Since it will probably be a web only thing, I think a library is the way to go.

I would expect a Channel like API in the Lua side, and a EventEmitter + send on the JS side.

Note that some restrictions on what can be sent should be applied, I would limit it to booleans, strings, numbers and arrays to begin with and later on move to tables, deep nested tables, Data objects, and so on only if it's possible to implement. For more complex stuff you could even use JSON as a common format...

I think this should be used only for simple stuff, like triggering a fetch in JavaScript from the Lua side and then pushing the result from JS to Lua or something like that.

**Comment by TannerRogalsky** [comment]
@Positive07 If you're interested, you should take a look at [other Lua-JS projects](https://github.com/daurnimator/lua.vm.js). It should show what's already been done and where the limitations are.

**Comment by Positive07** [comment]
Cool so you can import the complete JS global environment! Do you think it would be possible to incorporate that into love.js as a library?

**Comment by TannerRogalsky** [comment]
It's probably possible but the lack of finalizers in JS make it a less than idiomatic option for interop. There are probably cleaner options for this particular project than a full JS FFI for Lua.

**Comment by Positive07** [comment]
Oh I see, so a stripped down version like the one I mentioned would actually fit better right?

**Comment by TannerRogalsky** [comment]
It wouldn't suffer from the finalizers issue, at least.

**Comment by ndarilek** [comment]
Not sure to what extent this is a meaningful question, but are there any current ways to call out to JS functions from Lua modules loaded via Love.js?

I have a module that uses the ffi library to call out to a handful of OS-specific C functions in native libraries. The JS solution doesn't necessarily need to use the ffi library--I'd just need a way to detect that the game is running on the web, and to perform JS-specific calls in my handful of functions.

I know it's possible to call into JS via C macros, but I don't know that I can embed those in FFI code. Maybe there's some way to write and load a C wrapper, but I'm not sure how I'd package and run that in the Lua instance hosting my game, or indeed if I even can.
